sycl/source/detail/scheduler/circular_buffer_extended.cpp

```diff
@@ -0,0 +1,133 @@
+//==-- circular_buffer_extended.hpp - Circular buffer with host accessor ---==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <detail/scheduler/circular_buffer_extended.hpp>
+#include <detail/scheduler/scheduler.hpp>
+
+#include <algorithm>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+static inline bool doOverlap(const Requirement *LHS, const Requirement *RHS) {
+  size_t LHSStart = LHS->MOffsetInBytes;
+  size_t LHSEnd = LHSStart + LHS->MAccessRange.size() * LHS->MElemSize;
+
+  size_t RHSStart = RHS->MOffsetInBytes;
+  size_t RHSEnd = RHSStart + RHS->MAccessRange.size() * RHS->MElemSize;
+
+  if (LHSStart < RHSStart) {
+    return (RHSStart < LHSEnd) && (LHSEnd <= RHSEnd);
+  } else {
+    return (LHSStart < RHSEnd) && (RHSEnd <= LHSEnd);
+  }
+}
+
+static inline bool isHostAccessorCmd(Command *Cmd) {
+  return Cmd->getType() == Command::EMPTY_TASK &&
+         Cmd->MBlockReason == Command::BlockReason::HostAccessor;
+}
+
+size_t CircularBufferExtended::remove(value_type Cmd) {
+  if (!isHostAccessorCmd(Cmd)) {
+    auto NewEnd =
+        std::remove(MGenericCommands.begin(), MGenericCommands.end(), Cmd);
+    size_t RemovedCount = std::distance(NewEnd, MGenericCommands.end());
+    MGenericCommands.erase(NewEnd, MGenericCommands.end());
+
+    return RemovedCount;
+  }
+
+  // host accessor commands part
+  return eraseHostAccessorCommand(static_cast<EmptyCommand *>(Cmd));
+}
+
+void CircularBufferExtended::push_back(value_type Cmd, MemObjRecord *Record) {
+  // if EmptyCommand add to HA
+  if (isHostAccessorCmd(Cmd)) {
+    addHostAccessorCommand(static_cast<EmptyCommand *>(Cmd), Record);
+  } else
+    addGenericCommand(Cmd, Record);
+}
+
+std::vector<CircularBufferExtended::value_type>
+CircularBufferExtended::toVector() const {
+  std::vector<value_type> Result;
+  Result.reserve(MGenericCommands.size() + MHostAccessorCommandsXRef.size());
+
+  Result.insert(Result.end(), MGenericCommands.begin(), MGenericCommands.end());
+
+  for (EmptyCommand *Cmd : MHostAccessorCommands)
+    Result.push_back(Cmd);
+
+  return Result;
+}
+
+void CircularBufferExtended::addHostAccessorCommand(EmptyCommand *Cmd,
+                                                    MemObjRecord *Record) {
+  // 1. find the oldest command with doOverlap() = true amongst the List
+  //      => OldCmd
+  HostAccessorCommandSingleXRefT OldCmdIt;
+
+  if (Cmd->getRequirement()->MAccessMode == cl::sycl::access::mode::read)
+    OldCmdIt = MHostAccessorCommands.end();
```

Comment:
Could you please add a comment here to clarify why read accessors are handled in this way? The comment above seems to imply that the check for overlap is applied universally.